src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,446 @@
+        {
+            if (left.Equals(right, StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+            else
+            {
+                var normalizedLeft = NormalizeOSName(left);
+                var normalizedRight = NormalizeOSName(right);
+                return string.Equals(normalizedLeft, normalizedRight, StringComparison.OrdinalIgnoreCase);
+            }
+        }
+
+        private static string NormalizeOSName(string name)
+        {
+            if (name.Equals(OSX, StringComparison.OrdinalIgnoreCase))
+                return macOS;
+            return name;
+        }
+
+        private static DiagnosticDescriptor SelectRule(PlatformAttributeType attributeType)
+        {
+            if (attributeType == PlatformAttributeType.MinimumOSPlatformAttribute)
+                return MinimumOsRule;
+            if (attributeType == PlatformAttributeType.ObsoletedInOSPlatformAttribute)
+                return ObsoleteRule;
+            return RemovedRule;
+        }
+
+        private static AttributeData[] FindAllPlatformAttributesApplied(ImmutableArray<AttributeData> immediateAttributes, INamedTypeSymbol parent)
+        {
+            using var builder = ArrayBuilder<AttributeData>.GetInstance();
+            foreach (AttributeData attribute in immediateAttributes)
+            {
+                if (s_osPlatformAttributes.Contains(attribute.AttributeClass.Name))
+                {
+                    builder.Add(attribute);
+                }
+            }
+
+            while (parent != null)
+            {
+                var current = parent.GetAttributes();
+                foreach (var attribute in current)
+                {
+                    if (s_osPlatformAttributes.Contains(attribute.AttributeClass.Name) &&
+                        TargetPlatformAttribute != attribute.AttributeClass.Name)
+                    {
+                        builder.Add(attribute);
+                    }
+                }
+                parent = parent.BaseType;
+            }
+            return builder.ToArray();
+        }
+
+        private static bool IsSuppressedByAttribute(PlatformAttributeInfo diagnosingAttribute, ISymbol containingSymbol)
+        {
+            while (containingSymbol != null)
+            {
+                var attributes = containingSymbol.GetAttributes();
+                if (attributes != null)
+                {
+                    foreach (AttributeData attribute in attributes)
+                    {
+                        if (diagnosingAttribute.AttributeType.ToString() == attribute.AttributeClass.Name)
+                        {
+                            if (PlatformAttributeInfo.TryParsePlatformAttributeInfo(attribute, out PlatformAttributeInfo parsedAttribute))
+                            {
+                                if (IsOSPlatformsEqual(diagnosingAttribute.PlatformName, parsedAttribute.PlatformName) &&
```

Comment:
Should this be returning false if this innermost check fails? i.e. say if a method is tagged with (platform: A, min version: 2) and its containing type is tagged with (platform: A, min version: 1), and if we are invoking this method from a path where runtime checks only guarantee (platform: A and version >= 1), we should return false, but will return true here.

Reply:
No we need to check all attributes, a type can have many attributes and suppressing attribute can be at any level (member/class/assembly)

Reply:
But we should return false when even one of AttributeVersionsMatch check returns false after platforms matched, correct? Currently, that does not seem to be the case.

Reply:
Yes, that is done below outside of the while loop, after checking all attributes

Reply:
I meant the below will likely not work, I am not sure if that is considered a valid attribute annotation/scenario though:

```csharp
[MinimumOSPlatform("Windows11")]
class C
{
    [MinimumOSPlatform("Windows10")]
    void Windows10Method()
    {
        Windows11Method(); // This should be flagged, will not be flagged with current algorithm
    }

    [MinimumOSPlatform("Windows11")]
    void Windows11Method()
    {
    }
}
```

Reply:
No that `Windows11Method` call is suppressed by containing class attribute, but that might not your point, anyway the commented method is removed now
```csharp
[MinimumOSPlatform("Windows11")]
class C
{
...
```

Reply:
Sorry I meant to not have the attribute on the class. Let’s talk offline on this.

Reply:
Sure but this method removed, the logic changed completely please re-review the PR